sdk/dialer.go

```
@@ -0,0 +1,216 @@
+package sdk
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	proto "go.sia.tech/core/rhp/v4"
+	"go.sia.tech/core/types"
+	"go.sia.tech/coreutils/chain"
+	"go.sia.tech/coreutils/rhp/v4"
+	"go.sia.tech/coreutils/rhp/v4/quic"
+	"go.sia.tech/coreutils/rhp/v4/siamux"
+	"go.sia.tech/indexd/api"
+)
+
+var _ HostDialer = (*Dialer)(nil)
+
+// Dialer implements the HostDialer interface.
+type Dialer struct {
+	mu sync.Mutex
+
+	c      AppClient
+	appKey types.PrivateKey
+
+	addrs          map[types.PublicKey][]chain.NetAddress
+	conns          map[types.PublicKey]rhp.TransportClient
+	cachedSettings map[types.PublicKey]proto.HostSettings
+}
+
+// NewDialer returns a new Dialer.
+func NewDialer(c AppClient, appKey types.PrivateKey) *Dialer {
+	return &Dialer{
+		c:      c,
+		appKey: appKey,
+
+		addrs:          make(map[types.PublicKey][]chain.NetAddress),
+		conns:          make(map[types.PublicKey]rhp.TransportClient),
+		cachedSettings: make(map[types.PublicKey]proto.HostSettings),
+	}
+}
+
+// Close closes all the open connections on the dialer.
+func (d *Dialer) Close() error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	for _, conn := range d.conns {
+		if err := conn.Close(); err != nil {
+			return err
```

We probably want to collect errors here and continue closing the rest.

---

Just ignore any errors. They are unlikely to matter.